### PR TITLE
feat: add/port initial x/oracle types

### PR DESCRIPTION
**Pull request by justyntemme**
## Description
Add type definitions to oracle module

Issue: https://github.com/umee-network/umee/issues/97

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/oracle/types/assets.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+const (
+	MicroLunaDenom  = "uluna"
+	MicroUSDDenom   = "uusd"
+	MicroKRWDenom   = "ukrw"
+	MicroSDRDenom   = "usdr"
+	MicroCNYDenom   = "ucny"
+	MicroJPYDenom   = "ujpy"
+	MicroEURDenom   = "ueur"
+	MicroGBPDenom   = "ugbp"
+	MicroMNTDenom   = "umnt"
```
Comment (alexanderbez) [inline x/oracle/types/assets.go:13]: Remove what we don't need. We probably only need `MicroAtomDenom` and `MicroUmeeDenom`.

File: x/oracle/types/assets.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package types
+
+//nolint
```
Comment (alexanderbez) [inline x/oracle/types/assets.go:3]:
```suggestion
// Denomination constants
```

File: x/oracle/types/ballot.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+import (
+	"fmt"
+	"math"
+	"strconv"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// NOTE: we don't need to implement proto interface on this file
+//       these are not used in store or rpc response
```
Comment (alexanderbez) [inline x/oracle/types/ballot.go:12]:
```suggestion
```

File: x/oracle/types/ballot.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+// ExchangeRateBallot is a convenience wrapper around a ExchangeRateVote slice
+type ExchangeRateBallot []VoteForTally
+
+// ToMap return organized exchange rate map by validator
+func (pb ExchangeRateBallot) ToMap() map[string]sdk.Dec {
+	exchangeRateMap := make(map[string]sdk.Dec)
+	for _, vote := range pb {
+		if vote.ExchangeRate.IsPositive() {
+			exchangeRateMap[string(vote.Voter)] = vote.ExchangeRate
```
Comment (alexanderbez) [inline x/oracle/types/ballot.go:40]:
```suggestion
			exchangeRateMap[vote.Voter.String()] = vote.ExchangeRate
```

File: x/oracle/types/ballot.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+		}
+	}
+
+	return exchangeRateMap
+}
+
+// ToCrossRate return cross_rate(base/exchange_rate) ballot
```
Comment (alexanderbez) [inline x/oracle/types/ballot.go:47]:
```suggestion
// ToCrossRate return cross_rate(base/exchange_rate) ballot.
```

File: x/oracle/types/ballot.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+		Denom:        denom,
+		Voter:        voter,
+		Power:        power,
+	}
+}
+
+// ExchangeRateBallot is a convenience wrapper around a ExchangeRateVote slice
+type ExchangeRateBallot []VoteForTally
+
+// ToMap return organized exchange rate map by validator
```
Comment (alexanderbez) [inline x/oracle/types/ballot.go:35]:
```suggestion
// ToMap return organized exchange rate map by validator.
```

File: x/oracle/types/ballot.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+// VoteForTally is a convenience wrapper to reduce redundant lookup cost
+type VoteForTally struct {
+	Denom        string
+	ExchangeRate sdk.Dec
+	Voter        sdk.ValAddress
+	Power        int64
+}
+
+// NewVoteForTally returns a new VoteForTally instance
```
Comment (alexanderbez) [inline x/oracle/types/ballot.go:22]:
```suggestion
// NewVoteForTally returns a new VoteForTally instance.
```

File: x/oracle/types/ballot.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+	"math"
+	"strconv"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// NOTE: we don't need to implement proto interface on this file
+//       these are not used in store or rpc response
+
+// VoteForTally is a convenience wrapper to reduce redundant lookup cost
```
Comment (alexanderbez) [inline x/oracle/types/ballot.go:14]:
```suggestion
// VoteForTally is a convenience wrapper to reduce redundant lookup cost.
```

File: x/oracle/types/ballot.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+func NewVoteForTally(rate sdk.Dec, denom string, voter sdk.ValAddress, power int64) VoteForTally {
+	return VoteForTally{
+		ExchangeRate: rate,
+		Denom:        denom,
+		Voter:        voter,
+		Power:        power,
+	}
+}
+
+// ExchangeRateBallot is a convenience wrapper around a ExchangeRateVote slice
```
Comment (alexanderbez) [inline x/oracle/types/ballot.go:32]:
```suggestion
// ExchangeRateBallot is a convenience wrapper around a ExchangeRateVote slice.
```

File: x/oracle/types/ballot.go [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+	return exchangeRateMap
+}
+
+// ToCrossRate return cross_rate(base/exchange_rate) ballot
+func (pb ExchangeRateBallot) ToCrossRate(bases map[string]sdk.Dec) (cb ExchangeRateBallot) {
+	for i := range pb {
+		vote := pb[i]
+
+		if exchangeRateRT, ok := bases[string(vote.Voter)]; ok && vote.ExchangeRate.IsPositive() {
```
Comment (alexanderbez) [inline x/oracle/types/ballot.go:52]:
```suggestion
		if exchangeRateRT, ok := bases[vote.Voter.String()]; ok && vote.ExchangeRate.IsPositive() {
```

File: x/oracle/types/ballot.go [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+			vote.Power = 0
+		}
+
+		cb = append(cb, vote)
+	}
+
+	return
+}
+
+// Power returns the total amount of voting power in the ballot
```
Comment (alexanderbez) [inline x/oracle/types/ballot.go:66]:
```suggestion
// Power returns the total amount of voting power in the ballot.
```

File: x/oracle/types/ballot.go [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+func (pb ExchangeRateBallot) Power() int64 {
+	totalPower := int64(0)
```
Comment (alexanderbez) [inline x/oracle/types/ballot.go:68]:
```suggestion
	var totalPower int64
```

File: x/oracle/types/ballot.go [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+
+	return totalPower
+}
+
+// WeightedMedian returns the median weighted by the power of the ExchangeRateVote.
+// CONTRACT: ballot must be sorted
+func (pb ExchangeRateBallot) WeightedMedian() sdk.Dec {
+	totalPower := pb.Power()
+	if pb.Len() > 0 {
+		pivot := int64(0)
```
Comment (alexanderbez) [inline x/oracle/types/ballot.go:81]:
```suggestion
		var pivot int64
```

File: x/oracle/types/ballot.go [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+		for _, v := range pb {
+			votePower := v.Power
+
+			pivot += votePower
+			if pivot >= (totalPower / 2) {
+				return v.ExchangeRate
+			}
+		}
+	}
+	return sdk.ZeroDec()
```
Comment (alexanderbez) [inline x/oracle/types/ballot.go:91]:
```suggestion

	return sdk.ZeroDec()
```

File: x/oracle/types/ballot.go [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+
+	sum := sdk.ZeroDec()
+	for _, v := range pb {
+		deviation := v.ExchangeRate.Sub(median)
+		sum = sum.Add(deviation.Mul(deviation))
+	}
+
+	variance := sum.QuoInt64(int64(len(pb)))
+
+	floatNum, _ := strconv.ParseFloat(variance.String(), 64)
```
Comment (alexanderbez) [inline x/oracle/types/ballot.go:110]: nit: we should probably bubble up the error to the caller.

File: x/oracle/types/assets.go (hunk 2 above)
Comment (justyntemme) [reply x/oracle/types/assets.go]: I have saved with the auto-format, as well as used gofmt -s , and go fmt -w, and still receive the error.

File: x/oracle/types/ballot.go (hunk 15 above)
Comment (justyntemme) [reply x/oracle/types/ballot.go]: Fixed, not a nitpick at all, that could have been a bug in the future without err entries, great find!

File: x/oracle/types/codec.go [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+var (
+	amino = codec.NewLegacyAmino()
+
+	// ModuleCdc references the global x/oracle module codec. Note, the codec should
+	// ONLY be used in certain instances of tests and for JSON encoding as Amino is
+	// still used for that purpose.
+	//
+	// The actual codec used for serialization should be provided to x/staking and
```
Comment (toteki) [inline x/oracle/types/codec.go:37]: Is this comment about x/staking correct?

File: x/oracle/types/assets.go [hunk 17]
```diff
@@ -0,0 +1,15 @@
+//nolint
+// Denomination constants
```
Comment (alexanderbez) [inline x/oracle/types/assets.go]: This is incorrect. I've pushed a fix though.

**Comment by toteki** [comment]
Looks good

**Pull request merged**